### PR TITLE
Renewal trace

### PR DESCRIPTION
**Pull request by ybiquitous**
What?
-----

- Added `file` field to `ci_config` entry.
- Added `duration_in_ms` field to `message` entry.
- Added `truncated` field to some types' entry.
- Output total elapsed time.

Why?
----

- To improve the log page on Sider UI.

### CODE REVIEW

File: lib/runners/trace_writer.rb [hunk 1]
```diff
@@ -7,88 +7,99 @@ def initialize(writer:, sensitive_strings: [])
...
+      recorded_at = object[:recorded_at]
+      object = object.merge(recorded_at: recorded_at.utc.iso8601(3)) if recorded_at
       writer << Schema::Trace.anything.coerce(object)
     end
 
     private
 
+    def now
+      Time.now
+    end
```
Comment (ybiquitous) [inline lib/runners/trace_writer.rb]: [note] To make stubbing easier on tests.

File: sig/polyfill.rbi [hunk 2]
```diff
@@ -45,7 +45,7 @@ class Time
   def iso8601: -> String
   def strftime: (String) -> String
   def -: (Integer) -> self
-       | (Time) -> Integer
+       | (Time) -> Float
```
Comment (ybiquitous) [inline sig/polyfill.rbi]: http://ruby-doc.org/core-2.6.5/Time.html#method-i-2D

image

File: lib/runners/trace_writer.rb [hunk 3]
```diff
@@ -7,88 +7,99 @@ def initialize(writer:, sensitive_strings: [])
...
-    def command_line(command_line, recorded_at: Time.now)
-      self << { trace: 'command_line', command_line: command_line.map { |v| masked_string(v) }, recorded_at: recorded_at.utc.iso8601 }
+    def command_line(command_line, recorded_at: now)
+      self << { trace: :command_line, command_line: command_line.map { |v| masked_string(v) }, recorded_at: recorded_at }
     end
 
-    def status(status, recorded_at: Time.now)
-      self << { trace: 'status', status: status.exitstatus, recorded_at: recorded_at.utc.iso8601 }
+    def status(status, recorded_at: now)
+      self << { trace: :status, status: status.exitstatus, recorded_at: recorded_at }
```
Comment (ybiquitous) [inline lib/runners/trace_writer.rb]: > __[Querly]__ Process#exitstatus may be nil ([see on Sider](https://sider.review/gh/repos/202486306/pulls/624#issue-4988075))
>
> <details>
>
> Process#exitstatus is nil when the process does not exit.
> This means the process is abort(2)-ed or coredumped.
>
> Check Process#exited? to make sure exitstatus is a Integer.
>
> </details>

We are checking it via StrongJSON.

File: lib/runners/trace_writer.rb [hunk 4]
```diff
@@ -7,88 +7,99 @@ def initialize(writer:, sensitive_strings: [])
...
+      if block_given
+        start = now
         yield.tap do
-          duration = Time.now - start
-          self << { trace: 'message', message: "    -> #{format_duration_in_secs(duration)}s", recorded_at: (recorded_at + duration).utc.iso8601 }
+          duration = now - start
+          duration_in_ms = (duration * 1000).truncate
+          duration_in_ms = 1 if duration_in_ms == 0 # NOTE: Prevent zero (no time)
+          self << { trace: :message, message: "-> #{duration_in_ms / 1000.0}s", recorded_at: recorded_at + duration,
+                    truncated: false, duration_in_ms: duration_in_ms }
```
Comment (ybiquitous) [inline lib/runners/trace_writer.rb]: [note] When passing a block to `#message` method:

1. Set `duration_in_ms: 0` to the start entry.
2. Set `duration_in_ms: actual_duration` to the finish entry.

When not passing a block:

- Not set `duration_in_ms`.

**Comment by ybiquitous** [comment]
@yykamei Could you please review this (tomorrow is OK 😉 )?

File: lib/runners/trace_writer.rb [hunk 5]
```diff
@@ -7,88 +7,99 @@ def initialize(writer:, sensitive_strings: [])
...
+    end
+
     def each_slice(string, size:)
       string = string.dup
 
       while string.length > 0
         slice = string.slice!(0, size)
         if slice.length == size
-          yield(slice + "\\")
+          yield slice + "\\", true
```
Comment (yykamei) [inline lib/runners/trace_writer.rb]: Can we remove the suffix of `"\\"`?

Comment (ybiquitous) [reply lib/runners/trace_writer.rb]: Looks nice. If need, we can add it on front-end. I'll remove it 👍

Comment (ybiquitous) [reply lib/runners/trace_writer.rb]: Fixed via eee57649d135af9fe247b1f0b3555aed88133270

**Comment by ybiquitous** [comment]
Thank you!

**Pull request merged**